### PR TITLE
Enhancing metadata API to return upsert partition to primary key count map for both controller and server APIs

### PR DESCRIPTION
**Pull request by 9aman**
## Scope of the PR
Enhance **server and controller API** to return **partition to primary key count** for realtime tables that have **upsert** enabled:
1. Server API: returns the map for all the partitions hosted by the instance.
2. Controller API: returns the map for the table.

## Testing
The changes were tested locally by invoking the **server endpoint** followed by **controller endpoint** to test that there is no duplication in counting PK's. The replication of segments across servers could be a potential reason for duplication, but the controller API has taken care of it.

- Server responses: 
```
{
    "tableName": "upsertMeetupRsvp_REALTIME",
    "diskSizeInBytes": 0,
    "numSegments": 1,
    "numRows": 0,
    "columnLengthMap": {},
    "columnCardinalityMap": {},
    "maxNumMultiValuesMap": {},
    "columnIndexSizeMap": {},
    "upsertPartitionToPrimaryKeyCountMap": {
        "0": 48
    }
},


{
    "tableName": "upsertMeetupRsvp_REALTIME",
    "diskSizeInBytes": 0,
    "numSegments": 1,
    "numRows": 0,
    "columnLengthMap": {},
    "columnCardinalityMap": {},
    "maxNumMultiValuesMap": {},
    "columnIndexSizeMap": {},
    "upsertPartitionToPrimaryKeyCountMap": {
        "1": 45
    }
}
```

- Controller response

```
{
        "tableName": "upsertMeetupRsvp_REALTIME",
        "diskSizeInBytes": 0,
        "numSegments": 2,
        "numRows": 0,
        "columnLengthMap": {},
        "columnCardinalityMap": {},
        "maxNumMultiValuesMap": {},
        "columnIndexSizeMap": {},
        "upsertPartitionToPrimaryKeyCountMap": {
          "0": 48,
          "1": 45
        }
}

```

## Edited Response:

```
{
  "tableName": "upsertJsonMeetupRsvp_REALTIME",
  "diskSizeInBytes": 83788,
  "numSegments": 1,
  "numRows": 72,
  "columnLengthMap": {},
  "columnCardinalityMap": {},
  "maxNumMultiValuesMap": {},
  "columnIndexSizeMap": {},
  "upsertPartitionToServerPrimaryKeyCountMap": {
    "0": {
      "Server_100.84.215.146_7052": 95
    },
    "1": {
      "Server_100.84.215.146_7053": 90
    }
  }
}
```

image

### CODE REVIEW

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java [hunk 1]
```diff
@@ -283,9 +284,17 @@ public String getSegmentMetadata(
...
+    if (tableDataManager instanceof RealtimeTableDataManager) {
+      RealtimeTableDataManager realtimeTableDataManager = (RealtimeTableDataManager) tableDataManager;
+      upsertPartitionToPrimaryKeyCountMap = realtimeTableDataManager.getUpsertPartitionToPrimaryKeyCount();
+    }
+
     TableMetadataInfo tableMetadataInfo =
         new TableMetadataInfo(tableDataManager.getTableName(), totalSegmentSizeBytes, segmentDataManagers.size(),
-            totalNumRows, columnLengthMap, columnCardinalityMap, maxNumMultiValuesMap, columnIndexSizesMap);
+            totalNumRows, columnLengthMap, columnCardinalityMap, maxNumMultiValuesMap, columnIndexSizesMap,
+            upsertPartitionToPrimaryKeyCountMap);
```
Comment (ege-st) [inline pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:297]: Is there anything to worry about with compatibility by adding a new field to the response message?

**Comment by xiangfu0** [comment]
What if there is discrepancy for partitions with multiple replicas?

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 2]
```diff
@@ -140,16 +143,17 @@ public TableMetadataInfo getAggregatedTableMetadataFromServer(String tableNameWi
...
-    // Since table segments may have multiple replicas, divide diskSizeInBytes, numRows and numSegments by numReplica
-    // to avoid double counting, for columnAvgLengthMap, columnAvgCardinalityMap and maxNumMultiValuesMap, dividing by
-    // numReplica is not needed since totalNumSegments already contains replicas.
+    // Since table segments may have multiple replicas, divide diskSizeInBytes, numRows, numSegments and primary key
+    // count by numReplica to avoid double counting, for columnAvgLengthMap, columnAvgCardinalityMap and
+    // maxNumMultiValuesMap, dividing by numReplica is not needed since totalNumSegments already contains replicas.
     totalDiskSizeInBytes /= numReplica;
     totalNumSegments /= numReplica;
     totalNumRows /= numReplica;
+    upsertPartitionToPrimaryKeyCountMap.replaceAll((k, v) -> v / numReplica);
```
Comment (ege-st) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:152]: I think (and the preceding calculations for totalDiskSizeInBytes, etc.) this assumes that each replica will have the exact same number of PKs as the other replicas, but what happens if that is not the case (ie, one replica is inconsistent and has a different number of PKs)? That discrepancy will get lost here but could be very valuable to have.

Is it possible for `upsertPartitionToPrimaryKeyCountMap` to store PK count for each Replica of each partition/segment?

File: pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java [hunk 3]
```diff
@@ -703,6 +704,18 @@ public TableUpsertMetadataManager getTableUpsertMetadataManager() {
...
+  /**
+   * Retrieves a mapping of partition id to the primary key count for the partition.
+   *
+   * @return A {@code Map} where keys are partition id and values are count of primary keys for that specific partition.
+   */
+  public Map<Integer, Long> getUpsertPartitionToPrimaryKeyCount() {
+    if (isUpsertEnabled()) {
+      return _tableUpsertMetadataManager.getPartitionToPrimaryKeyCount();
+    }
+    return new HashMap<>();
```
Comment (klsince) [inline pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java:716]: Collections.emptyMap()

File: pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/models/DummyTableUpsertMetadataManager.java [hunk 4]
```diff
@@ -72,6 +73,11 @@ public PartitionUpsertMetadataManager getOrCreatePartitionManager(int partitionI
   public void stop() {
   }
 
+  @Override
+  public Map<Integer, Long> getPartitionToPrimaryKeyCount() {
+    return null;
```
Comment (klsince) [inline pinot-integration-tests/src/test/java/org/apache/pinot/integration/tests/models/DummyTableUpsertMetadataManager.java:78]: if `null` is allowed, then can return null for RealtimeTableDataManager.getUpsertPartitionToPrimaryKeyCount() as well. And annotate the method with `@Nullable`

File: pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapTableUpsertMetadataManager.java [hunk 5]
```diff
@@ -45,6 +46,15 @@ public void stop() {
     }
   }
 
+  @Override
+  public Map<Integer, Long> getPartitionToPrimaryKeyCount() {
+    Map<Integer, Long> partitionToPrimaryKeyCount = new HashMap<>();
+    for (Integer partitionID : _partitionMetadataManagerMap.keySet()) {
```
Comment (klsince) [inline pinot-segment-local/src/main/java/org/apache/pinot/segment/local/upsert/ConcurrentMapTableUpsertMetadataManager.java:52]: can do _partitionMetadataManagerMap.forEach()

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java [hunk 6]
```diff
@@ -140,16 +143,17 @@ public TableMetadataInfo getAggregatedTableMetadataFromServer(String tableNameWi
       return v;
     });
 
-    // Since table segments may have multiple replicas, divide diskSizeInBytes, numRows and numSegments by numReplica
-    // to avoid double counting, for columnAvgLengthMap, columnAvgCardinalityMap and maxNumMultiValuesMap, dividing by
-    // numReplica is not needed since totalNumSegments already contains replicas.
+    // Since table segments may have multiple replicas, divide diskSizeInBytes, numRows, numSegments and primary key
+    // count by numReplica to avoid double counting, for columnAvgLengthMap, columnAvgCardinalityMap and
+    // maxNumMultiValuesMap, dividing by numReplica is not needed since totalNumSegments already contains replicas.
```
Comment (klsince) [inline pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:148]: how about keeping all partition replicas' count in the map? then, we can find out any discrepancy with this API

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java (hunk 2 above)
Comment (9aman) [reply pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java:152]: @klsince @ege-st I agree to above point. Adding replica-level information can help improve the usability of the API.

I can return partition to server to PK count Map 

{
"0": {
            "Server1":50,
            "Server2:28,
        },
},

**Comment by soumitra-st** [comment]
Why we are limiting the scope to upsert enabled realtime table vs all realtime tables?

File: pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java [hunk 7]
```diff
@@ -283,9 +284,23 @@ public String getSegmentMetadata(
       }
     }
 
+    // fetch partition to primary key count for realtime tables that have upsert enabled
+    Map<Integer, Long> upsertPartitionToPrimaryKeyCountMap = new HashMap<>();
+    if (tableDataManager instanceof RealtimeTableDataManager) {
+      RealtimeTableDataManager realtimeTableDataManager = (RealtimeTableDataManager) tableDataManager;
+      upsertPartitionToPrimaryKeyCountMap = realtimeTableDataManager.getUpsertPartitionToPrimaryKeyCount();
```
[the remark this answers is not in the document]
Comment (ege-st) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:291]: My concern here is: how will a user tell the difference between a table that does _not_ have Upsert and a table that _has_ Upsert but no rows (is empty)?

[the remark this answers is not in the document]
Comment (KKcorps) [reply pinot-server/src/main/java/org/apache/pinot/server/api/resources/TablesResource.java:291]: Good callout. IMO, if upsert is not enabled we should return -1 and not 0

**Review by KKcorps — APPROVED** [review]

File: pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java (hunk 2 above)
Comment (ege-st) [reply pinot-controller/src/main/java/org/apache/pinot/controller/util/ServerSegmentMetadataReader.java]: I think this would be great.  Thanks.

**Pull request merged**